modules/currency-l1/src/main/scala/org/tessellation/currency/l1/http/DataApplicationRoutes.scala

```diff
@@ -47,64 +51,136 @@ final case class DataApplicationRoutes[F[_]: Async: Hasher: SecurityProvider: L1
+          .flatMap { transactions =>
+            validateDataTransactions(transactions) {
+              transactions.traverse { signedTransaction =>
+                signedTransaction.value match {
+                  case dataUpdate: DataUpdate =>
+                    Signed(dataUpdate, signedTransaction.proofs)
+                      .toHashedWithSignatureCheckEnsuringSerializer[F](dataApplication.serializeUpdate _)
+                      .asInstanceOf[F[Either[InvalidSignatureForHash[DataTransaction], Hashed[DataTransaction]]]]
+
+                  case feeTransaction: FeeTransaction =>
+                    Signed(feeTransaction, signedTransaction.proofs)
+                      .toHashedWithSignatureCheckEnsuringSerializer[F](serializeFeeTransaction[F] _)
+                      .asInstanceOf[F[Either[InvalidSignatureForHash[DataTransaction], Hashed[DataTransaction]]]]
+                }
+              }.flatMap { results =>
+                val errors = results.collect { case Left(_) => InvalidSignature.toApplicationError }
+                val validTransactions = results.collect { case Right(hashed) => hashed }
+
+                if (errors.nonEmpty) {
+                  BadRequest("Invalid signature in data transactions")
+                } else if (validTransactions.isEmpty) {
+                  BadRequest("No valid data transactions found")
+                } else {
+                  val (dataUpdateHashes, feeTransactionHashes) = validTransactions
+                    .partition(_.signed.value.isInstanceOf[DataUpdate])
+
+                  val response =
+                    ("dataUpdateHashes" ->> dataUpdateHashes.map(_.hash.value).mkString(", ")) ::
+                      ("feeTransactionHashes" ->> feeTransactionHashes.map(_.hash.value).mkString(", ")) ::
+                      HNil
+
+                  dataTransactionsQueue.offer(NonEmptyList.fromListUnsafe(validTransactions.map(_.signed))) >>
+                    Ok(response)
+                }
               }
+            }
           }
-        }
-        .handleUnknownError
+      }
+
+      def handleSingleDataUpdate: F[Response[F]] = {
+        implicit val decoder: EntityDecoder[F, Signed[DataUpdate]] = dataApplication.signedDataEntityDecoder
+        req
+          .as[Signed[DataUpdate]]
+          .flatMap { signedUpdate =>
+            validateDataTransactions(NonEmptyList.of(signedUpdate)) {
+              signedUpdate
+                .toHashedWithSignatureCheckEnsuringSerializer[F](dataApplication.serializeUpdate _)
+                .flatMap {
+                  case Left(_) => BadRequest(InvalidSignature.toApplicationError)
+                  case Right(hashed) =>
+                    val dataTransaction: NonEmptyList[Signed[DataTransaction]] =
+                      NonEmptyList.of(hashed.signed)
+                    dataTransactionsQueue.offer(dataTransaction) >>
+                      Ok(("hash" ->> hashed.hash.value) :: HNil)
+                }
+            }
```

Comment:
I think it should reuse the "handleMultiple" process. Just the response output would be different to be backward compatible, but the processing should be no different I believe. Hence, you can trigger "handleMultiple" here and map the output to focus on a single hash instead of a list of hashes (just get head element).

Reply:
Updating